File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        Time: O(n), where n is the no. of nodes
        Space: O(n)

        We just go level by level and save the nodes in the same level in a subset array
        """
        if root is None:
            return None

        output = []
        q = deque([root])
        while q:
            subset = []
            for _ in range(len(q)):
                node = q.pop()
                subset.append(node.val)
                if node.left: q.appendleft(node.left)
                if node.right: q.appendleft(node.right)
            output.append(subset)
        return output

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type Queue struct {
    list []TreeNode
}

func (q *Queue) dequeue () TreeNode {
    poppedValue := q.list[0]
    q.list = q.list[1:]
    return poppedValue
}

func (q *Queue) enqueue (node *TreeNode) {
    q.list = append(q.list, *node)
}

func levelOrder(root *TreeNode) [][]int {
    if root == nil {
        return nil
    } 
    
    var output [][]int
    q := &Queue{}
    q.enqueue(root)

    for len(q.list) > 0 {
        level := []int{}
        for range q.list {
            node := q.dequeue()
            level = append(level, node.Val)
            if node.Left != nil {
                q.enqueue(node.Left)
            }
            if node.Right != nil {
                q.enqueue(node.Right)
            }
        }
        output = append(output, level)
    }
    return output
}

